security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -322,12 +325,16 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
 				"error details %s\n", conIDresourceNamePrefix, discReq.Node.Id, firstRequestFlag, discReq.VersionInfo,
 				discReq.ErrorDetail)
 
-			// In ingress gateway agent mode, if the first SDS request is received but kubernetes secret is not ready,
-			// wait for secret before sending SDS response. If a kubernetes secret was deleted by operator, wait
-			// for a new kubernetes secret before sending SDS response.
-			if s.st.ShouldWaitForIngressGatewaySecret(conID, resourceName, token) {
+			// In ingress gateway agent mode, if the first SDS request is received but Ingress gateway secret which is
+			// setup as kubernetes secret is not ready, wait for secret before sending SDS response.
+			// If a kubernetes secret was deleted by operator, wait for a new kubernetes secret before sending SDS response.
+			// If workload uses file mounted certs i.e. "FILE_MOUNTED" is set to true, it does not depend on ingress gateway
```

Comment:
Hi @ramaraochavali - Where is "FILE_MOUNTED" set to true?

Reply:
BTW I assume we support both (file mounted and sds gateway) ?  and depends on how secrets/credentials are specified within the user's gateway resource, the code determines which path it should be used.

cc @howardjohn

Reply:
https://github.com/istio/istio/blob/7cb751f183999845ad58100589584c4084838463/pkg/istio-agent/sds-agent.go#L89 - It was added a while ago to pilot-agent (specically to sds-agent) to indicate a workload is using file mounted certs. Deployments which have that can set that env var

Reply:
Thank you @ramaraochavali !  I think this may be @lei-tang 's question as well.

Reply:
With the existing ingress gateway secret mechanism, why the mechanism of file mounted certs is needed for ingress gateway? What is the pros and cons of the existing ingress gateway secret mechanism versus the proposed file-mounted certs for ingress gateway?

Reply:
People have existing CAs/secret management that inject certs as files in the pod. Also VMs...

For the average case you would always use Secret reading - but if you have existing integrations it's simply not possible

Reply:
Correct. And additionally for sidecar with has `FILE_MOUNTED_CERTS` to true, this fails to loads those certs and send to proxy with this error
```
2020-05-20T10:14:43.765581Z	warn	sds	resource:file-root:/etc/identity/ca/cacerts.pem waiting for ingress gateway secret for proxy "sidecar~10.100.172.250~istio-ordering-98c49bb4c-q854l.service-mesh~service-mesh.svc.cluster.local"
```

Reply:
For non-ingress workloads, the mechanism of file mounted certs makes sense because no equivalent mechanisms exist for non-ingress workloads.
But for ingress gateways, the existing ingress gateway secret mechanism is equivalent to the proposed mechanism of file mounted certs. Adding an equivalent mechanism to the ingress gateway will make the code more complicated and may create confusions to the users.

Reply:
Lei, there are many many tools that integrate to provision certificate into pod. https://www.hashicorp.com/blog/injecting-vault-secrets-into-kubernetes-pods-via-a-sidecar/ is one example. I don't think it's reasonable to ask people to change their entire PKI infrastructure

Reply:
> But for ingress gateways, the existing ingress gateway secret mechanism is equivalent to the proposed mechanism of file mounted certs
It is actually not equivalent. Because @howardjohn  mentioned, in Salesforce (and I am sure lot of other enterprises will also have similar setup) we have an existing PKI infrastructure that manages cert delivery and rotation as file in PODS (for non-ingreess as well as ingress) which is security vetted.  I think we can not assume kubernetes secrets.,

Also apart from Ingress, this check/wait is also blocking sidecars to come up. What about VMs? What about deployments that do not use ingress gateway? In those cases this should not block sidecars to come up

Reply:
Adding more mechanisms to provision certificates into pods may help users. But to make the proposed mechanism truly useful to potential users, the PR should help users to understand the proposed mechanism, e.g., describing the differences between the proposed mechanism and the existing mechanisms, as well as their different usage scenarios. The review comments have repeatedly asked for such information and finally we see some answers in the comments.

Reply:
Sorry, it was not completely clear to me. I have added comment about usage scenario for file mounted certs. But I am happy to additional comments if you have any suggestions